src/front/glsl/builtins.rs

```diff
@@ -118,31 +138,83 @@ pub fn inject_builtin(declaration: &mut FunctionDeclaration, module: &mut Module
+                    "textureLodOffset" => MacroCall::Texture { proj: false, offset: true, level_type: TextureLevelType::Lod, },
+                    // textureOffset(gsampler, gvec+1 P, ivec offset, [float bias]);
+                    "textureOffset" => MacroCall::Texture { proj: false, offset: true, level_type: TextureLevelType::None, },
+                    // textureProj(gsampler, gvec+1 P, [float bias]);
+                    "textureProj" => MacroCall::Texture { proj: true, offset: false, level_type: TextureLevelType::None, },
+                    // textureProjGrad(gsampler, gvec+1 P, gvec dPdx, gvec dPdy);
+                    "textureProjGrad" => MacroCall::Texture { proj: true, offset: false, level_type: TextureLevelType::Grad(shadow), },
+                    // textureProjGradOffset(gsampler, gvec+1 P, gvec dPdx, gvec dPdy, ivec offset);
+                    "textureProjGradOffset" => MacroCall::Texture { proj: true, offset: true, level_type: TextureLevelType::Grad(shadow), },
+                    // textureProjLod(gsampler, gvec+1 P, float lod);
+                    "textureProjLod" => MacroCall::Texture { proj: true, offset: false, level_type: TextureLevelType::Lod, },
+                    // textureProjLodOffset(gsampler, gvec+1 P, gvec dPdx, gvec dPdy, ivec offset);
+                    "textureProjLodOffset" => MacroCall::Texture { proj: true, offset: true, level_type: TextureLevelType::Lod, },
+                    // textureProjOffset(gsampler, gvec+1 P, ivec offset, [float bias]);
+                    "textureProjOffset" => MacroCall::Texture { proj: true, offset: true, level_type: TextureLevelType::None, },
+                    _ => unreachable!(),
                 };
 
-                // Shadow, arrayed or both 3D images are not allowed
-                if (shadow || arrayed) && dim == 0b11
-                    || (builtin == MacroCall::TextureLod
-                        && (dim != 0b00 && shadow && arrayed || dim == 0b10 && shadow || !bias))
-                {
+                // Parse out the variant settings.
+                let proj = matches!(builtin, MacroCall::Texture { proj: true, .. });
+                let grad = matches!(builtin, MacroCall::Texture { level_type: TextureLevelType::Grad(_), .. });
+                let lod = matches!(builtin, MacroCall::Texture { level_type: TextureLevelType::Lod, .. });
+                let offset = matches!(builtin, MacroCall::Texture { offset: true, .. });
+
+                if variant {
+                    let supports_variant = proj && !shadow;
+                    if !supports_variant {
+                        continue;
+                    }
+                }
+
+                if bias {
+                    let supports_bias = matches!(builtin, MacroCall::Texture { level_type: TextureLevelType::None, .. }) && !shadow;
+                    if !supports_bias {
+                        continue;
+                    }
+                }
```

Comment:
nit: I don't like this nested `if`s you can do `variant && (proj && ! shadow)` or keep the `supports_variant` but move it out of the `if` (compilers will almost certainly inline this) and rust defines short circuiting so there's no additional computations if the first check doesn't pass (it's [defined](https://doc.rust-lang.org/reference/expressions/operator-expr.html#lazy-boolean-operators) at the language level so you can always expect it).